src/Npgsql/Internal/TypeHandling/TypeHandlerResolver.cs

```diff
@@ -5,23 +5,25 @@ namespace Npgsql.Internal.TypeHandling
     /// <summary>
     /// An Npgsql resolver for type handlers. Typically used by plugins to alter how Npgsql reads and writes values to PostgreSQL.
     /// </summary>
-    public interface ITypeHandlerResolver
+    public abstract class TypeHandlerResolver
     {
         /// <summary>
         /// Resolves a type handler given a PostgreSQL type name, corresponding to the typname column in the PostgreSQL pg_type catalog table.
         /// </summary>
         /// <remarks>See <see href="https://www.postgresql.org/docs/current/catalog-pg-type.html" />.</remarks>
-        NpgsqlTypeHandler? ResolveByDataTypeName(string typeName);
+        public abstract NpgsqlTypeHandler? ResolveByDataTypeName(string typeName);
 
         /// <summary>
         /// Resolves a type handler given a .NET CLR type.
         /// </summary>
-        NpgsqlTypeHandler? ResolveByClrType(Type type);
+        public abstract NpgsqlTypeHandler? ResolveByClrType(Type type);
+
+        public virtual NpgsqlTypeHandler? ResolveValueDependentValue(object value) => null;
```

Comment:
We really want to pass the type into this method as well. I'm always on guard around 'handler resolution' methods where the baseline is runtime types without a Type argument or generic method. Passing it in gives the call site control to pick the desired semantics (static/runtime). Fun things * always * happen with values with a type that derives from an expected type (even though this method is only used for structs right now). Let's just say I've been burned by MediatR/Newtonsoft/older automapper versions and others that failed to anticipate this in their api design.

Reply:
Speaking of structs, we're ok with boxing all datetimes?

Reply:
> ResolveByValue

The thing is that unlike the other Resolve methods, this should *only* resolve DateTime (or other value-dependent cases, if we end up having them). It's important for this to return null for anything else (e.g. int), so that ConnectorTypeMapper can flow down to ResolveByClrType - which does caching. This ensures that for all regular types, the second resolution is just one cache dictionary lookup from the 2nd time.

So it seems important to indicate this in the method name (naming is hard :tm:)

Reply:
> Speaking of structs, we're ok with boxing all datetimes?

I've already got a plan for this :) For regular non-generic NpgsqlParameter, it's already boxed anyway in NpsgqlParameter.Value. Generic `NpgsqlParameter<T>` resolves through `ResolveByValue<T>`, which should call into a corresponding generic method on the resolvers, where they'd do checks for the value types, which would get elided.. This would mean that value type resolution for generic NpgsqlParameter wouldn't even need to go through a dictionary lookup, which seems nice - I'll do this and benchmark too.